File: 08/lab_08.R
## setup
## 
library(pacman)
p_load(char = c('arrow','tidyverse','sf','tmap','tmaptools',
                'USAboundaries','tigris','rnaturalearth','osmar',
                'lubridate'))

accidents <- arrow::read_feather('data/accidents.feather')
accidents_19 <- accidents %>% 
  mutate(year = year(start_time),
         month = month(start_time),
         hour = hour(start_time)) %>% 
  filter(year==2019)

## Data munging

accidents_19_sf <- accidents_19 %>% 
  st_as_sf(coords = c('start_lng','start_lat'), 
           crs = 4326) %>% 
  rename(geom_acc = geometry)
us_map <- USAboundaries::us_states()

acc_for_maps <- accidents %>% 
  select(severity, start_time, start_lng, start_lat, 
         description, state, zipcode, timezone, weather_timestamp:weather_condition) %>% 
  left_join(us_map, by = c('state'='state_abbr'))

## Plotting slices

### Choropleth of number of accidents by state

accidents_19 %>% 
  count(state, month) %>% 
  group_by(state) %>% 
  mutate(prop = 100*n/sum(n)) %>% 
  ungroup() %>% 
  left_join(us_map, by=c('state'='state_abbr'))-> bl

ggplot(bl %>% filter(month %in% c(1,4,9,12)))+
  geom_sf(aes(geometry=geometry, fill=prop))+
  facet_wrap(~month, ncol=2)+
  scale_fill_viridis_c()

ggplot(bl %>% filter(month==6))+
  geom_sf(aes(geometry=geometry, fill=prop))

accidents_19 %>% count(state, severity) %>% 
  group_by(state) %>% 
  mutate(prop = n/sum(n)*100) %>% 
  ungroup() %>% 
  left_join(us_map, by=c('state'='state_abbr'))-> acc_severity

ggplot(acc_severity)+
  geom_sf(aes(geometry=geometry, fill = prop))+
  facet_wrap(~severity) +
  scale_fill_viridis_c()

accidents_19 %>% 
  filter(state=='MD') %>% 
  st_as_sf(coords = c('start_lng','start_lat'), 
           crs = 4326) -> acc_19_md
ggplot(acc_19_md)+
  geom_sf(data=us_counties(states='MD'), aes(geometry=geometry))+
  geom_sf(aes(geometry=geometry, color=severity, fill=severity), 
          size=0.2, alpha=0.4)


counties2 <- sf::st_read('~/Downloads/tl_2019_us_county.shp')

File: 08/lab_08.py
# ## Setup

# **Load packages**

import pandas as pd
import matplotlib.pyplot as plt
import geopandas as gp
import pyarrow
# %matplotlib inline

# **Load data**

accidents = pd.read_feather('data/accidents.feather')
abbrevs = pd.read_csv('data/state_abbrevs.csv')

us_states = gp.read_file('/Users/abhijit/Downloads/tl_2019_us_state.shp')
us_counties = gp.read_file('/Users/abhijit/Downloads/tl_2019_us_county.shp')


# **Data  munging**

us_counties = (us_counties
              .merge(us_states[['STATEFP','NAME']], how='left',
                    left_on='STATEFP',right_on='STATEFP')
              .rename(columns = {'NAME_x':'county_name','NAME_y':'state_name'}))
us_counties = us_counties.to_crs(epsg=4326)

accidents['year'] = pd.DatetimeIndex(accidents['start_time']).year
accidents['month'] = pd.DatetimeIndex(accidents['start_time']).month
accidents_19 = accidents[accidents.year==2019]
accidents.shape

# ## Plotting accidents by state

by_state = (accidents_19.state
     .value_counts().reset_index()
     .merge(abbrevs, how='left', left_on='index', right_on='Abbreviation')
     .merge(us_states, how='left', left_on='US.State', right_on='NAME')
     )

by_state = gp.GeoDataFrame(by_state)

by_state.plot();

by_state.plot(column='state');

by_state.plot(column='state',
             cmap='plasma');

by_state.plot(column='state',
             cmap='plasma',
             legend=True);

from mpl_toolkits.axes_grid1 import make_axes_locatable
fig, ax= plt.subplots(1,1)
divider = make_axes_locatable(ax)

cax = divider.append_axes('right', size='5%', pad=0.1)

by_state.plot(column='state',
             cmap='plasma',
             legend=True,
             ax=ax, 
             cax=cax);


fig, ax = plt.subplots(1,1)
by_state.plot(column='state',
             cmap='plasma',
             legend=True,
             ax = ax,
             legend_kwds={'label': "Accidents by state",
                         "orientation" : "horizontal"});


# ## Dot map

acc_19_md = accidents_19.query('state=="MD"')
acc_19_md = gp.GeoDataFrame(acc_19_md, 
                           geometry=gp.points_from_xy(acc_19_md.start_lng,
                                                      acc_19_md.start_lat,
                                                     crs=4326))

# _Using transparency_

fig, ax = plt.subplots(1,1)
acc_19_md.plot(ax = ax, column='severity', alpha=0.3, s = 1);
us_counties.query('state_name=="Maryland"').plot(ax=ax,color='white', 
                                                 edgecolor='black', alpha=0.2);


# _Changing layers_

fig, ax = plt.subplots(1,1)
us_counties.query('state_name=="Maryland"').plot(ax=ax,color='white', 
                                                 edgecolor='black')
acc_19_md.plot(ax = ax, column='severity', alpha=0.3, s = 1,
              cmap = 'YlOrBr',
               legend=True);


